test/UnitTests/MSTest.CoreAdapter.Unit.Tests/Execution/TestMethodInfoTests.cs

```diff
@@ -806,6 +806,40 @@ public void TestMethodInfoInvokeShouldCallDiposeForDisposableTestClass()
         Verify(disposeCalled);
     }
 
+#if NET6_0_OR_GREATER
+    public void TestMethodInfoInvokeShouldCallDiposeAsyncForAsyncDisposableTestClass()
+    {
+        var asyncDisposeCalled = false;
+        DummyTestClassWithAsyncDisposable.DisposeAsyncMethodBody = () => asyncDisposeCalled = true;
+        var ctorInfo = typeof(DummyTestClassWithAsyncDisposable).GetConstructors().Single();
+        var testClass = new TestClassInfo(typeof(DummyTestClassWithAsyncDisposable), ctorInfo, null, _classAttribute, _testAssemblyInfo);
+        var method = new TestMethodInfo(typeof(DummyTestClassWithAsyncDisposable).GetMethod("DummyTestMethod"), testClass, _testMethodOptions);
+
+        method.Invoke(null);
+
+        Verify(asyncDisposeCalled);
+    }
+
+    public void TestMethodInfoInvokeShouldCallAsyncDiposeThenDiposeForDisposableAsyncDisposableTestClass()
+    {
+        var order = 0;
+        var disposeCalledOrder = 0;
+        var disposeAsyncCalledOrder = 0;
+
+        DummyTestClassWithAsyncDisposableAndDisposable.DisposeMethodBody = () => disposeCalledOrder = ++order;
+        DummyTestClassWithAsyncDisposableAndDisposable.DisposeAsyncMethodBody = () => disposeAsyncCalledOrder = ++order;
+
+        var ctorInfo = typeof(DummyTestClassWithAsyncDisposableAndDisposable).GetConstructors().Single();
+        var testClass = new TestClassInfo(typeof(DummyTestClassWithAsyncDisposableAndDisposable), ctorInfo, null, _classAttribute, _testAssemblyInfo);
+        var method = new TestMethodInfo(typeof(DummyTestClassWithAsyncDisposableAndDisposable).GetMethod("DummyTestMethod"), testClass, _testMethodOptions);
+
+        method.Invoke(null);
+
+        Verify(disposeCalledOrder == 2);
+        Verify(disposeAsyncCalledOrder == 1);
```

Comment:
The test method is clear as-is but as it starts to have a couple lines of code, I want to use this as an example for you to use the AAA comments (Arrange, Act, Assert). The goal is to help identify which part of the test is there to "arrange"/"setup" the test to be in the right state, which part is the actual behavior you want to test and which part is the assertion of the behavior.

```suggestion
        // Arrange
        var order = 0;
        var disposeCalledOrder = 0;
        var disposeAsyncCalledOrder = 0;

        DummyTestClassWithAsyncDisposableAndDisposable.DisposeMethodBody = () => disposeCalledOrder = ++order;
        DummyTestClassWithAsyncDisposableAndDisposable.DisposeAsyncMethodBody = () => disposeAsyncCalledOrder = ++order;

        var ctorInfo = typeof(DummyTestClassWithAsyncDisposableAndDisposable).GetConstructors().Single();
        var testClass = new TestClassInfo(typeof(DummyTestClassWithAsyncDisposableAndDisposable), ctorInfo, null, _classAttribute, _testAssemblyInfo);
        var method = new TestMethodInfo(typeof(DummyTestClassWithAsyncDisposableAndDisposable).GetMethod("DummyTestMethod"), testClass, _testMethodOptions);

        // Act
        method.Invoke(null);

        // Assert
        Verify(disposeCalledOrder == 2);
        Verify(disposeAsyncCalledOrder == 1);
```